### DISCUSSION TITLE
Export of risk analysis error (missing translations)

### DISCUSSION THREAD

**Discussion by mrapnx** [monarc-project/MonarcAppFO#405] [General]
When I try to export the risk analysis with assessments
![ex1](https://user-images.githubusercontent.com/64850246/148732136-4a63ce52-c8ea-4b2d-a34a-a4d113fe41e1.PNG)

an error message is being displayed
![ex2](https://user-images.githubusercontent.com/64850246/148732030-22ea6506-a269-4979-8c64-3656eadaea2d.PNG)

 `Fehler beim Abrufen von api/client-anr/5/export: Call to a member function getKey() on null (500)`

Do you need any log files to analyse the problem further?

Kind regards
Marcus

**Comment by ruslanbaidan** [comment]
Hello Marcus,

We definitely need you console error response to be able to identify and provide you with a fix for this issue.
You could attach it here or send us an email.

Thank you.

**Reply by mrapnx** [reply]
Dear Ruslan,

no prob, please find all Firefox Console Outputs after Click on "Export" here:


```
XHRPOSThttp://MONARCHOST/api/client-anr/5/export
[HTTP/1.1 500 Internal Server Error 2015ms]

Possibly unhandled rejection: {"data":{"errors":[{"message":"Call to a member function getKey() on null","error":"error-exception","exception":{"class":"Error","file":"/var/lib/monarc/fo/vendor/monarc/core/src/Service/OperationalRiskScalesExportService.php","line":110,"message":"Call to a member function getKey() on null","stacktrace":"#0 /var/lib/monarc/fo/vendor/monarc/core/src/Service/OperationalRiskScalesExportService.php(53): Monarc\\Core\\Service\\OperationalRiskScalesExportService->getOperationalRiskScaleCommentData()\n#1 /var/lib/monarc/fo/vendor/monarc/core/src/Service/AnrService.php(426): Monarc\\Core\\Service\\OperationalRiskScalesExportService->generateExportArray()\n#2 /var/lib/monarc/fo/vendor/monarc/core/src/Service/AnrService.php(278): Monarc\\Core\\Service\\AnrService->generateExportArray()\n#3 /var/lib/monarc/fo/vendor/monarc/frontoffice/src/Controller/ApiAnrExportController.php(35): Monarc\\Core\\Service\\AnrService->exportAnr()\n#4 /var/lib/monarc/fo/vendor/laminas/laminas-mvc/src/Controller/AbstractRestfulController.php(467): Monarc\\FrontOffice\\Controller\\ApiAnrExportController->create()\n#5 /var/lib/monarc/fo/vendor/laminas/laminas-mvc/src/Controller/AbstractRestfulController.php(428): Laminas\\Mvc\\Controller\\AbstractRestfulController->processPostData()\n#6 /var/lib/monarc/fo/vendor/laminas/laminas-eventmanager/src/EventManager.php(331): Laminas\\Mvc\\Controller\\AbstractRestfulController->onDispatch()\n#7 /var/lib/monarc/fo/vendor/laminas/laminas-eventmanager/src/EventManager.php(188): Laminas\\EventManager\\EventManager->triggerListeners()\n#8 /var/lib/monarc/fo/vendor/laminas/laminas-mvc/src/Controller/AbstractController.php(103): Laminas\\EventManager\\EventManager->triggerEventUntil()\n#9 /var/lib/monarc/fo/vendor/laminas/laminas-mvc/src/Controller/AbstractRestfulController.php(312): Laminas\\Mvc\\Controller\\AbstractController->dispatch()\n#10 /var/lib/monarc/fo/vendor/laminas/laminas-mvc/src/DispatchListener.php(138): Laminas\\Mvc\\Controller\\AbstractRestfulController->dispatch()\n#11 /var/lib/monarc/fo/vendor/laminas/laminas-eventmanager/src/EventManager.php(331): Laminas\\Mvc\\DispatchListener->onDispatch()\n#12 /var/lib/monarc/fo/vendor/laminas/laminas-eventmanager/src/EventManager.php(188): Laminas\\EventManager\\EventManager->triggerListeners()\n#13 /var/lib/monarc/fo/vendor/laminas/laminas-mvc/src/Application.php(331): Laminas\\EventManager\\EventManager->triggerEventUntil()\n#14 /var/lib/monarc/fo/public/index.php(34): Laminas\\Mvc\\Application->run()\n#15 {main}"}}]},"status":500,"config":{"method":"POST","transformRequest":[null],"transformResponse":[null],"jsonpCallbackParam":"callback","url":"api/client-anr/5/export","data":{"id":5,"password":"","assessments":1,"methodSteps":false,"interviews":false,"controls":false,"recommendations":false,"soas":false,"records":false},"headers":{"Accept":"application/json, text/plain, */*","Content-Type":"application/json;charset=utf-8","token":"28f2dac2a1d4ef240af130e7e14e8fc25f782e7d93b52506b4d75bab67eedaf761ded1d51c78f4.35477443"},"cached":false},"statusText":"Internal Server Error","xhrStatus":"complete"} angular-common-libs.js:132:202
    Angular 46
Das Cookie “cf_28394_cta_92976” wurde als Cross-Site-Cookie gegen “http://MONARCHOST/api/client-anr/5/export” behandelt, da das Schema nicht übereinstimmt. export
Das Cookie “cf_28394_cta_92747” wurde als Cross-Site-Cookie gegen “http://MONARCHOST/api/client-anr/5/export” behandelt, da das Schema nicht übereinstimmt. export
Das Cookie “cf_28394_id” wurde als Cross-Site-Cookie gegen “http://MONARCHOST/api/client-anr/5/export” behandelt, da das Schema nicht übereinstimmt. export
Das Cookie “cf_28394_person_last_update” wurde als Cross-Site-Cookie gegen “http://MONARCHOST/api/client-anr/5/export” behandelt, da das Schema nicht übereinstimmt. 
```

Please note that I replaced the actual hostname on this log with _MONARCHOST_.

Kindest regards
Marcus

**Reply by ruslanbaidan** [reply]
Dear Marcus,

There is (are) missing translation for some comments.
The issue was fixed on BackOffice side and and in the migration. Now, in the models custom scales comments created with all the languages and empty string by default.

You could validate if you have any missing comments translations by running the following query:
`select c.id, c.`comment_translation_key`, t.* from operational_risks_scales_comments c left join translations t on c.comment_translation_key = t.translation_key where c.anr_id = {YOUR_ANR_ID}`
If starting from the 3rd column you see nullable values for some rows, then you need to create records for them.

Just execute an insert with empty string values or real comments if you know, something like this:
`INSERT INTO `translations` (`anr_id`, `type`, `translation_key`, `lang`, `value`, `creator`, `created_at`)
VALUES
	({YOUR_ANR_ID}, 'operational-risk-scale-comment', {YOUR_COMMENT_TRANSLATION_KEY_FROM_SQL_QUERY_RESULT}, 'de', '', 'admin@admin.localhost', NOW());`

**Reply by mrapnx** [reply]
Dear Ruslan,
> You could validate if you have any missing comments translations by running the following query: `select c.id, c.`comment_translation_key`, t.* from operational_risks_scales_comments c left join translations t on c.comment_translation_key = t.translation_key where c.anr_id = **{YOUR_ANR_ID}**` If starting from the 3rd column you see nullable values for some rows, then you need to create records for them.

How can I identify "my" ANR_ID? Is this 
`{"id":5,"password":"","assessments":1,"methodSteps":false,"interviews":false,"controls":false,"recommendations":false,"soas":false,"records":false} 
`from the "data" parameter of api/client-anr/5/export, which would be "5" ?

Kindest regards
Marcus

**Comment by ruslanbaidan** [comment]
Hello Marcus,

anr_id  (analysis ID) is indicated in your browser URL, when you open a risk analysis.
You can see here:
![Screenshot 2022-01-17 at 19 19 56](https://user-images.githubusercontent.com/3246171/149821159-f940a5aa-c589-4f53-9232-1de94d992a8a.png)

**Comment by uwocgh** [comment]
Dear Ruslan,

my colleague Marcus handed that task over to me, and I created these entries in the translation table. But unfortunately the error is still there. 
Fehler beim Abrufen von api/client-anr/1/export: Call to a member function getKey() on null (500)

Any suggestions what we can do in this case?

Regards,
Uwe

**Reply by ruslanbaidan** [reply]
Hello Uwe,

You can also check with the following query, if all scale types are translated:
`select c.id, c.label_translation_key, t.* from operational_risks_scales_types c left join translations t on c.label_translation_key = t.translation_key where c.anr_id = {YOUR_ANR_ID}`

It is important to be sure that there are all comments and types have translations to the language that was used for the analysis creation.

**Comment by uwocgh** [comment]
Hi Ruslan,

thx that looks better.

Regards,
Uwe
